include/deal.II/base/array_view.h

```diff
@@ -0,0 +1,219 @@
+//
+// Copyright (C) 2004 - 2015 by the deal.II authors
+//
+// This file is part of the deal.II library.
+//
+// The deal.II library is free software; you can use it, redistribute
+// it, and/or modify it under the terms of the GNU Lesser General
+// Public License as published by the Free Software Foundation; either
+// version 2.1 of the License, or (at your option) any later version.
+// The full text of the license can be found in the file LICENSE at
+// the top level of the deal.II distribution.
+//
+// ---------------------------------------------------------------------
+
+#ifndef dealii__array_view_h
+#define dealii__array_view_h
+
+#include <deal.II/base/config.h>
+#include <deal.II/base/exceptions.h>
+#include <boost/type_traits/remove_cv.hpp>
+#include <boost/type_traits/is_same.hpp>
+#include <boost/utility/enable_if.hpp>
+
+#include <vector>
+
+DEAL_II_NAMESPACE_OPEN
+
+
+/**
+ * A class that represents a window of memory locations of type
+ * @p ElementType and presents it as if it was an array that
+ * can be accessed via an <code>operator[]</code>. In essence,
+ * this class is nothing more than just a pointer to the first
+ * location and an integer that represents the length of the array
+ * in elements.
+ *
+ * The advantage of using this class is that you don't have to pass
+ * around pairs of pointers and that <code>operator[]</code> checks
+ * for the validity of the index with which you subscript this
+ * array view.
+ *
+ * This class can handle views to both non-constant and constant memory
+ * locations. If you want to represent a view of a constant array,
+ * then the template argument type of this class needs to be
+ * @p const as well. The following code snippet gives an example:
+ * @code
+ *   std::vector<int>       array       = get_data();  // a writable array
+ *
+ *   ArrayView<int> view (&array[5], 5);               // a view of elements 5..9 (inclusive)
+ *   view[2] = 42;                                     // array[7] is set to 42
+ *
+ *   ArrayView<const int> const_view (&array[5], 5);   // same view, but read-only
+ *   int element_7 = const_view[2];                    // returns 42
+ *   const_view[2] = 42;                               // error, can't write into this view
+ * @endcode
+ *
+ * @ingroup data
+ * @author Wolfgang Bangerth, 2015
+ */
+template <typename ElementType>
+class ArrayView
+{
+public:
+  /**
+   * A typedef that denotes the "value_type" of this container-like
+   * class, i.e., the type of the element it "stores" or points to.
+   */
+  typedef ElementType value_type;
+
+  /**
+   * Constructor.
+   * @param[in] starting_element A pointer to the first element of the array
+   *   this object should represent.
+   * @param[in] n_elements The length (in elements) of the chunk of memory
+   *   this object should represent.
+   */
+  ArrayView (value_type       *starting_element,
+             const std::size_t n_elements);
```

Comment:
This class cannot determine that the memory locations `starting_element[0]` to `starting_element[n_elements-1]` are actually valid but relies on the user who creates the view. This is opposed to the index checking within the array. In the `VectorView` class we have a statement of the form:

> Notice that NO CHECKS are performed on the actual memory, and if you try to access illegal areas of memory, your computer will suffer from it.

Something similar should be stated here. It is easy to confuse the safety regarding the index access by the safety of the given input arguments.